server/src/main/java/org/elasticsearch/injection/Injector.java

```diff
@@ -0,0 +1,396 @@
+        if (constructors.size() == 1) {
+            return constructors.get(0);
+        }
+        var injectConstructors = constructors.stream().filter(c -> c.isAnnotationPresent(Inject.class)).toList();
+        if (injectConstructors.size() == 1) {
+            return injectConstructors.get(0);
+        }
+        LOGGER.trace("No suitable constructor for {}", type);
+        return null;
+    }
+
+    /**
+     * When creating <code>specsByClass</code>, we compute a kind of "inheritance closure"
+     * in the sense that, for each class <code>C</code>, we not only add an entry for <code>C</code>,
+     * but we also add {@link AliasSpec} entries for all abstract supertypes.
+     * <p>
+     * This method is part of the recursion that achieves this.
+     */
+    private static void aliasSuperinterfaces(Class<?> classToScan, Class<?> classToAlias, Map<Class<?>, InjectionSpec> specsByClass) {
+        for (var i : classToScan.getInterfaces()) {
+            registerSpec(new AliasSpec(i, classToAlias), specsByClass);
+            aliasSuperinterfaces(i, classToAlias, specsByClass);
+        }
+    }
+
+    private static void registerSpec(InjectionSpec spec, Map<Class<?>, InjectionSpec> specsByClass) {
+        Class<?> requestedType = spec.requestedType();
+        var existing = specsByClass.put(requestedType, spec);
+        if (existing == null || existing.equals(spec)) {
+            LOGGER.trace("Register spec: {}", spec);
+        } else {
+            AmbiguousSpec ambiguousSpec = new AmbiguousSpec(requestedType, spec, existing);
+            LOGGER.trace("Ambiguity discovered for {}", requestedType);
+            specsByClass.put(requestedType, ambiguousSpec);
+        }
+    }
+
+    private List<InjectionStep> injectionPlan(Set<Class<?>> requiredClasses, Map<Class<?>, InjectionSpec> specsByClass) {
+        LOGGER.trace("Constructing instantiation plan");
+        Set<Class<?>> allParameterTypes = new HashSet<>();
+        specsByClass.values().forEach(spec -> {
+            if (spec instanceof MethodHandleSpec m) {
+                m.parameters().stream().map(ParameterSpec::injectableType).forEachOrdered(allParameterTypes::add);
+            }
+        });
+
+        var plan = new Planner(specsByClass, requiredClasses, allParameterTypes).injectionPlan();
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("Injection plan: {}", plan.stream().map(Object::toString).collect(joining("\n\t", "\n\t", "")));
+        }
+        return plan;
+    }
+
+    /**
+     * <em>Evolution note</em>: there may be cases in the where we allow the user to
+     * supply a {@link java.lang.invoke.MethodHandles.Lookup} for convenience,
+     * so that they aren't required to make things public just to participate in injection.
+     */
+    private static MethodHandles.Lookup lookup() {
+        return MethodHandles.publicLookup();
+    }
+
+    private static final Logger LOGGER = LogManager.getLogger(Injector.class);
```

Comment:
see other examples of logger usage in ES, these are normally lowercase, and declared at the top of the file, let's follow that pattern throughout this PR